experiments_main.cpp

```diff
@@ -211,75 +233,38 @@ struct SeqGenModule {
     }
 
     void compute_pairwise_dists() {
-        int num_seqs = seqs.size();
-        if (FLAGS_mutation_pattern == "pairs") {
-            dists = new3D<double>(8, num_seqs, 1, -1);
-            pbar_start(seqs.size() / 2);
-#pragma omp parallel for default(shared)
-            for (size_t i = 0; i < seqs.size(); i += 2) {
-                int j = i + 1;
-                dists[0][i][0] = edit_distance(seqs[i], seqs[j]);
-                dists[1][i][0] = hamming_dist(mh_sketch[i], mh_sketch[j]);
-                dists[2][i][0] = hamming_dist(wmh_sketch[i], wmh_sketch[j]);
-                dists[3][i][0] = hamming_dist(omh_sketch[i], omh_sketch[j]);
-                dists[4][i][0] = l1_dist(ten_sketch[i], ten_sketch[j]);
-                dists[5][i][0] = l1_dist2D_minlen(slide_sketch[i], slide_sketch[j]);
-                pbar_inc();
-            }
-        } else {
-            dists = new3D<double>(8, num_seqs, num_seqs, 0);
-            pbar_start(seqs.size());
-#pragma omp parallel for default(shared)
-            for (size_t i = 0; i < seqs.size(); i++) {
-                for (size_t j = i + 1; j < seqs.size(); j++) {
-                    dists[0][i][j] = edit_distance(seqs[i], seqs[j]);
-                    dists[1][i][j] = hamming_dist(mh_sketch[i], mh_sketch[j]);
-                    dists[2][i][j] = hamming_dist(wmh_sketch[i], wmh_sketch[j]);
-                    dists[3][i][j] = hamming_dist(omh_sketch[i], omh_sketch[j]);
-                    dists[4][i][j] = l1_dist(ten_sketch[i], ten_sketch[j]);
-                    dists[5][i][j] = l1_dist2D_minlen(slide_sketch[i], slide_sketch[j]);
+        for (size_t g =0; g < seqs.size(); g += FLAGS_group_size ) { // g: group offset
+            for (size_t i = 0; i < FLAGS_group_size ; i++) {
+                for (size_t j = i + 1; j < FLAGS_group_size && g + j < seqs.size(); j++) {
+                    pairs.push_back({ g + i, g + j }); // g+i, g+i -> index i & j in the group
                 }
-                pbar_inc();
             }
         }
+        dists = new2D<double>(6, pairs.size());
+        PB_init(pairs.size());
+#pragma omp parallel for default(shared)
+        for (size_t pi=0; pi< pairs.size(); pi++ ) {
+            size_t si = pairs[pi].first, sj = pairs[pi].second;
+            dists[0][pi]=(edit_distance(seqs[si], seqs[sj]));
+            dists[1][pi]=(hamming_dist(mh_sketch[si], mh_sketch[sj]));
+            dists[2][pi]=(hamming_dist(wmh_sketch[si], wmh_sketch[sj]));
+            dists[3][pi]=(hamming_dist(omh_sketch[si], omh_sketch[sj]));
+            dists[4][pi]=(l1_dist(ten_sketch[si], ten_sketch[sj]));
+            dists[5][pi]=(l1_dist2D_minlen(slide_sketch[si], slide_sketch[sj]));
```

Comment:
It would be cool if we could refactor this such that all these assignments can be replaced by a for loop.

Maybe all sketch types could implement a common interface?

(That would imply virtual function calls, but I don't really see a way around that apart from function pointers which is more ugly.)